controllers/datasource_controller.go

```diff
@@ -401,3 +392,22 @@ func (r *GrafanaDatasourceReconciler) SetupWithManager(mgr ctrl.Manager, ctx con
 
 	return err
 }
+
+func (r *GrafanaDatasourceReconciler) GetMatchingDatasourceInstances(ctx context.Context, datasource *v1beta1.GrafanaDatasource, k8sClient client.Client) (v1beta1.GrafanaList, error) {
+	instances, err := GetMatchingInstances(ctx, k8sClient, datasource.Spec.InstanceSelector)
+	if err != nil || len(instances.Items) == 0 {
+		datasource.Status.NoMatchingInstances = true
+		err = r.Client.Status().Update(ctx, datasource)
```

Comment:
I think this will be a bit inconsistent in its behavior:
`err` generated by `GetMatchingInstances` is never returned from the function as it gets overridden by the status update. So, successful status update erases the error while failing status update makes the caller think as if there are no matching instances.
The question is whether in this helper you want to return status update errors at all or better to just log it while returning only errors directly related to the matching instances lookup.

I haven't thought it out fully, but if we always consider having no matching instances as an error, maybe it'll make sense to embed this logic into `GetMatchingInstances` (e.g. define a custom error for that and return it).

Reply:
Yeah I was thinking about this as well.
I guess one of the questions we have to ask our self is why wouldn't the operator be able to update the status? Is it because it was removed or access rights? Is there any other reasonable reason?

But as you say it's probably better to give the user the correct information that there actually is an hit.
We probably should go over to custom errors but it's probably something that we should do everywhere.
So lets do it in a separate PR and see if we can get this merged if it looks okay with you?

Reply:
The code still overwrites the error. For status updates errors, we need to either use another variable name or use the syntax that scopes err to the if structure, e.g.:

```go
func (r *GrafanaDashboardReconciler) GetMatchingDashboardInstances(ctx context.Context, dashboard *v1beta1.GrafanaDashboard, k8sClient client.Client) (v1beta1.GrafanaList, error) {
	instances, err := GetMatchingInstances(ctx, k8sClient, dashboard.Spec.InstanceSelector)
	if err != nil || len(instances.Items) == 0 {
		dashboard.Status.NoMatchingInstances = true
		if err := r.Client.Status().Update(ctx, dashboard); err != nil {
			r.Log.Info("unable to update the status of %v, in %v", dashboard.Name, dashboard.Namespace)
		}
		return v1beta1.GrafanaList{}, err
	}
	dashboard.Status.NoMatchingInstances = false
	if err := r.Client.Status().Update(ctx, dashboard); err != nil {
		r.Log.Info("unable to update the status of %v, in %v", dashboard.Name, dashboard.Namespace)
	}

	return instances, err
}
```

Reply:
Thanks, my brain wasn't really working...
I use the code you suggested, thanks @weisdd